### DISCUSSION TITLE
Связь многие ко многим

### DISCUSSION THREAD

**Discussion by juramebel** [totumonline/totum-ru-issues-and-discussions#153] [Как это сделать?] [answered]
Здравствуйте. 
Есть задача, не могу придумать ,как решить ее красиво.
Есть справочник номенклатуры.
Определенные позиции должны быть связаны друг с другом. Причем это могут быть разные отношения. 
Например. Есть плитный материал. Ему могут соответствовать кромочные материалы разных производителей. 
Условно это связь кромка - плита. 
Есть стяжки, в которых определенные виды эксцентрика соответствуют определенным болтам.
Можно ли придумать какой то простую связь? 
Пока делаю так - есть отдельное поле, определяющее тип связи, в которой участвует единица. И в зависимости от это выбирается селект в другом поле. 
Но это как то не универсально. И одна единица не может участвовать более чем в одном типе связи. 
Может какая то промежуточная таблица, в которой записываются пары в два колонки?
Но как отслеживать тогда удаление связи... или ее изменение...

**Comment by ttm-alexeypolunin** [comment]
Я тут не совсем понял:

+ Плита — Кромки
+ Эксцентрик — Болты

Если например в Заказе вам нужно показать определенные Кромки к выбранной Плите, то где-то должно быть указано, какие Кромки к этой Плите подходят.

Например в таблице с Плитами мы делаем поле в котором выбираем конкретные Кромки или тип Кромки.

Там где мы выбираем (в Заказе) например для Кромок:

```
=: selectRowListForSelect(table: 'table_with_items'; field: 'name'; where: 'id' = $sel)
sel: select(table: 'table_with_items'; field: 'avaliable'; where: 'id' = #source_item)
```

**Reply by juramebel** [reply]
"Если например в Заказе вам нужно показать определенные Кромки к выбранной Плите, то где-то должно быть указано, какие Кромки к этой Плите подходят."

Верно. И сейчас это просто поле селекта с множественным значением. Селект по этой же таблице.

Но это подходит только для одного типа связи.

**Reply by ttm-alexeypolunin** [reply]
Ок. Вот к строке Плита 1 выбрали подходящие Кромки. Что будет являться вторым типом связи для Плита 1?

**Reply by juramebel** [reply]
Ну это скорее из теории. 
Типа вдруг потом понадобится, а я уже все сделал не так.
Но, например, тип связи аналог.

**Reply by ttm-alexeypolunin** [reply]
Ну вот есть в теории «многие ко многим», что люди под этим в конкретном решении понимают я не догоняю :(

Можете рассказать что вы сделали, тк я саму задачу не понимаю.

Если под этим понимается, что у нас несколько Плит могут сочетаться с одними и теми-же кромками — то это все тот-же множественный Селект. В одной Плите одни Кромки в другой другие, но им ничего не мешает пересекаться. Из теории вроде как на этом все.

**Reply by juramebel** [reply]
Может быть связь сочетаемых кромок. 
Может быть аналоги материала.
И я это решаю через множественный селект и отдельное поле тип связи. 
А идея была можно ли сделать просто связь. Без типа. 
Например ЛДСП 1 связано с кромкой 2.
Также ЛДСП 1 связано с ЛДСП 2.
Без типа. 
Но эта связь видна также и со стороны ЛДСП 2. 
И кромки 2.
То есть связь двусторонняя. 
И можно делать временные таблицы, показывая связанные материалы. 
И использовать в превью селектора при занесении спецификаций. 
Плохо сформулировал. 
Попробую попозже потолковее. 
Как это организовать не совсем понимаю.

**Reply by ttm-alexeypolunin** [reply]
Те есть: `ЛДСП_1` <-> `ЛДСП_2` <— `[Кромка 1, Кромка 2]` и хочется, что бы при выборе `ЛДСП_1` мы видели не только `ЛДСП_2`, но и связанные с ней Кромки?

**Reply by juramebel** [reply]
Это прям граф какой то)
Если так будет, будет непонятна связь типа подходящие. 
Нет, наверное все проще. 
Нужно условно говоря, если я выбрал в селекте какие то позиции, то у них в селекте появилась бы ссылка на позицию из которой я выбираю.
Но если написать код обновления такой по изменению поля селекта, я получу зацикливание, как понимаю.

**Reply by ttm-alexeypolunin** [reply]
А, ок — это делается через `set`, напишу завтра пример.

**Reply by ttm-alexeypolunin** [reply]
Вот код для экшена по изменению:

```
= : listReplace(list: $list; action: $set; value: "value")

	list: listMinus(list: $#nfv; list: $#onfv)

set: set(table: $#ntn; field: $#nf + $minus; where: 'id' = $#value; log: true)

	minus: listMinus(list: json`[#id]`; list: $sel)
		sel: select(table: $#ntn; field: $#nf; where: 'id' = $#value)
```
этот код не учитывает никаким образом, что можно снять значение — это потом доделаю.

**Reply by juramebel** [reply]
Сходу ничего не понял, но это работает)
Спасибо!
Буду ждать вторую часть

**Comment by ttm-alexeypolunin** [comment, marked as answer]
**UPD:** Нужно условно говоря, если я выбрал в селекте какие то позиции, то у них в селекте появилась бы ссылка на позицию из которой я выбираю.

**Вот полный код — ставите все три триггера на добавление, удаление и изменение:**

```
// processing add and change

a1= : listReplace(list: $list; action: $set; value: "value")

	list: listMinus(list: $#nfv; list: $#onfv)

set: set(table: $#ntn; field: $#nf + $minus; where: 'id' = $#value; log: true)

	minus: listMinus(list: json`[#id]`; list: $sel)
		sel: select(table: $#ntn; field: $#nf; where: 'id' = $#value)
		
// processing when we turn off some items when change field and when we will delete row

a2=: listReplace(list: $deleted_items; action: $set_ch; value: "value_change")

	deleted_items: listMinus(list: $#onfv; list: $#nfv)
	
set_ch: set(table: $#ntn; field: $#nf - #old.id; where: 'id' = $#value_change; log: true)
```

**Reply by konditerov** [reply]
этот код можно использовать для тегов?
Смысл такой, что бы статусы сделать тегами и по мере исполнения - убирать теги из таблицы, таблица данная видна нескольким юзерам и каждый с ней работает отмечая выполненное, тем самым удаляются теги=статусы.
?

**Reply by ttm-alexeypolunin** [reply]
Я  не совсем понял вашу задачу, можете ее в отдельной ветке описать — мне кажется там все в 10 раз проще будет.
